src/Raven.Server/Documents/Sharding/Operations/FetchDocumentsFromShardsOperation.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Raven.Client.Documents.Commands;
+using Raven.Client.Http;
+using Raven.Server.Documents.Sharding.Handlers;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Utils;
+
+namespace Raven.Server.Documents.Sharding.Operations
+{
+    public readonly struct FetchDocumentsFromShardsOperation : IShardedReadOperation<GetDocumentsResult, GetShardedDocumentsResult>
+    {
+        private readonly TransactionOperationContext _context;
+        private readonly ShardedDocumentHandler _handler;
+        private readonly ShardedDatabaseContext _databaseContext;
+        private readonly Dictionary<int, ShardLocator.IdsByShard<string>> _idsByShards;
+        private readonly int _size;
+        private readonly string _etag;
+        private readonly string[] _includePaths;
+        private readonly bool _metadataOnly;
+
+
+        public FetchDocumentsFromShardsOperation(
+            TransactionOperationContext context,
+            ShardedDocumentHandler handler,
+            Dictionary<int, ShardLocator.IdsByShard<string>> idsByShards,
+            int size,
+            string etag,
+            string[] includePaths,
+            bool metadataOnly)
+        {
+            _context = context;
+            _handler = handler;
+            _databaseContext = handler.DatabaseContext;
+            _idsByShards = idsByShards;
+            _size = size;
+            _etag = etag;
+            _includePaths = includePaths;
+            _metadataOnly = metadataOnly;
+        }
+
+        public string ExpectedEtag => _etag;
+
+        public GetShardedDocumentsResult CombineResults(Memory<GetDocumentsResult> results)
+        {
+            var span = results.Span;
+            var docs = new BlittableJsonReaderObject[_size];
+            var includesMap = new Dictionary<string, BlittableJsonReaderObject>(StringComparer.OrdinalIgnoreCase);
+            var missingIncludes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            for (var shardNumber = 0; shardNumber < span.Length; shardNumber++)
+            {
+                var cmd = span[shardNumber];
+                if (cmd == null)
+                    continue;
+
+                var cmdResults = cmd.Results;
+
+                var cmdIncludes = cmd.Includes;
+                if (cmdIncludes != null)
+                {
+                    BlittableJsonReaderObject.PropertyDetails prop = default;
+                    for (int i = 0; i < cmdIncludes.Count; i++)
+                    {
+                        cmdIncludes.GetPropertyByIndex(i, ref prop);
+                        includesMap[prop.Name] = ((BlittableJsonReaderObject)prop.Value).Clone(_context);
+                    }
+                }
+
+                var positions = _idsByShards.ElementAt(shardNumber).Value.Positions;
```

Comment:
Huh? You are accessing a dictionary by numeric index?

Reply:
removed